replica/table.cc

```diff
@@ -820,7 +838,7 @@ sstables::compaction_type_options::split tablet_storage_group_manager::split_com
 future<> tablet_storage_group_manager::split_all_storage_groups() {
     sstables::compaction_type_options::split opt = split_compaction_options();
 
-    for (auto& storage_group : storage_groups()) {
+    for (auto& storage_group : storage_groups() | boost::adaptors::map_values) {
```

Comment:
Please note, that with dynamic groups allocation we would need to get back to iterating over all possible (`tablet_count()`) storage groups here. Otherwise, if the loop preempts, iterators may get invalidated. (I think that's the easiest way to prevent it)

Other loops over storage groups don't preempt.

Reply:
I think we should allocate and deallocate storage_groups only via update_effective_replication_map, and so we can use a rwlock to protect storage_groups() while it's scanned.

Reply:
I see. Currently storage groups are deallocated in tablet_cleanup (cg needs to be stopped there and we can't keep it after it's stopped), but it doesn't change much I think

Reply:
This should be called with effective_replication_map_ptr held, so no changes can happen during iteration. Please verify that.

Reply:
Maybe I'm missing something but I do not see where it could be held

Reply:
It must be held. How can we split storage groups if someone else can move a new one in or an old one out?

Reply:
Practically any call here should be with the effective replication map locked, one way or another. It's just too fragile otherwise.

Reply:
if erm is taken after pending updates reached replica, those updates could still be applied. we're thinking about using a rwlock to synchronize changes to erm and scanning of storage groups, that reflect current erm.

for example:

1) split and migration start concurrently
2) migration advances to allow_write...
3) split loop starts and take erm, and preempt
3) migration update reaches replica and causes replica table to allocate a new storage group

maybe preemptive loops over storage groups should hold erm in addition to the rwlock.

Reply:
actually, I think we shouldn't hold erm during split_all_groups(). otherwise, we lose the ability to split groups concurrently to migration. rwlock should be enough, right?. when allocating / deallocating / splitting groups, we'll synchronize with any preemptive loops over groups through rwlock.

Reply:
and splitting groups can take long, blocking topology changes for a significant amount of time. let's say a shard has 1T of data, with compaction at 30M/s/shard. let's assume 100G intersecting (100 tablets) with split point. That's almost 1 hour.

Reply:
Currently the preparation work for splitting is not not changing topology.
Only when all tablets are split, we apply the topology change and "promote" the split compaction groups to be first-class storage groups.  The is done via update_effective_replication_map that must be atomic with respect the to storage group manager.  Therefore tablet mugration sode effects of adding or removing tablet ownership should be applied using the same path and the storage group manager should update its storage groups based on the delta between the current erm and the new erm.

Reply:
Cc @Deexie ☝️